pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -112,6 +122,107 @@ func (c *Controller) EnqueueObj(obj interface{}) {
 	c.queue.Add(podKey)
 }
 
+func (c *Controller) checkDeletedSvc(obj interface{}) (*corev1.Service, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
+	}
+	svc, ok := deletedState.Obj.(*corev1.Service)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Service: %v", deletedState.Obj)
+	}
+	return svc, nil
+}
+
+func (c *Controller) EnqueueSvcUpdate(oldObj, newObj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both set of pods (corresponding to old and new selector) need to be considered
+	newSvc, isSvc := newObj.(*corev1.Service)
+	if !isSvc {
+		var err error
+		newSvc, err = c.checkDeletedSvc(newObj)
+		if err != nil {
+			klog.Errorf("Got error while processing event update: %v", err)
+			return
+		}
+	}
+
+	oldSvc := oldObj.(*corev1.Service)
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion {
+		return
+	}
+
+	podKeys := append(c.getPodsFromService(oldSvc), c.getPodsFromService(newSvc)...)
+	for _, podKey := range podKeys {
+		c.queue.Add(podKey)
+	}
+}
+
+func (c *Controller) EnqueueSvc(obj interface{}) {
+	svc, isSvc := obj.(*corev1.Service)
+	if !isSvc {
+		var err error
+		svc, err = c.checkDeletedSvc(obj)
+		if err != nil {
+			klog.Errorf("Got error while processing event update: %v", err)
+			return
+		}
+	}
+
+	for _, podKey := range c.getPodsFromService(svc) {
+		c.queue.Add(podKey)
+	}
+}
+
+func (c *Controller) getPodsFromService(svc *corev1.Service) []string {
+	var pods []string
+	annotations := svc.GetAnnotations()
+	if val, ok := annotations[NPLServiceAnnotation]; !ok || (ok && val != "true") {
+		return pods
+	}
+
+	appSelectorValue, ok := svc.Spec.Selector["app"]
```

Comment:
Is it right? We are not matching Pods in the standard way but requiring the service must have "app" key as a selector and only checking it but not other expressions?

Reply:
Could you expand on this a bit, I might be missing something. Isn't defining the pod selector in a service using this scheme the standard way? It assumes  that the service has
```
spec:
  selector:
    app: MyApp
```
and the pods would have 
```
labels:
  app: MyApp
```

I'm essentially following this
https://kubernetes.io/docs/concepts/services-networking/service/ 
> This specification creates a new Service object named "my-service", which targets TCP port 9376 on any Pod with the app=MyApp label.

Reply:
what Quan means is that there is nothing special about the `app` label. It's just an example in the K8s documentation. I can define Services using a completely different set of labels. For example I can use `foo: bar` as my label selector.

Reply:
so you need to consider the entire selector